src/EmbedClientEx.ts

```diff
@@ -0,0 +1,746 @@
+          'number'
+        ) {
+          this._client.send('session:tokens', {
+            api_token: undefined,
+            api_token_ttl: undefined,
+            navigation_token: undefined,
+            navigation_token_ttl: undefined,
+            session_reference_token_ttl: undefined,
+          })
+          return
+        }
+
+        // Session has expired. Need to send valid tokens for backwards
+        // compatibility
+        if (cookielessSession.cookielessSessionReferenceTokenTtl === 0) {
+          this._client.send('session:tokens', {
+            api_token: 'session expired',
+            api_token_ttl: 240,
+            navigation_token: 'session expired',
+            navigation_token_ttl: 240,
+            session_reference_token_ttl: 0,
+          })
+          return
+        }
+
+        if (updateGenerateTokensTime) {
+          // At this point we have valid tokens. Use the same tokens until
+          // 2 minutes before expiry time. This prevents multiple IFRAMEs
+          // from hammering the generate token endpoint
+          if (cookielessSession.cookielessSessionReferenceTokenTtl === 0) {
+            this._cookielessSessionExpired = true
+            cookielessSession.generateTokensTime = 0
+          } else {
+            let newGenerateTokensTime = Date.now()
+            const {
+              cookielessSessionReferenceTokenTtl,
+              cookielessApiTokenTtl,
+              cookielessNavigationTokenTtl,
+            } = cookielessSession
+            const ttl = Math.min(
+              cookielessSessionReferenceTokenTtl,
+              Number(cookielessApiTokenTtl),
+              Number(cookielessNavigationTokenTtl)
+            )
+            if (ttl < 60) {
+              newGenerateTokensTime = Date.now()
+            } else if (ttl < 120) {
+              newGenerateTokensTime = Date.now() + 60 * 1000
+            } else {
+              newGenerateTokensTime = Date.now() + (ttl - 120) * 1000
+            }
+            cookielessSession.generateTokensTime = newGenerateTokensTime
+          }
+        }
+
+        // Send good tokens
+        this._client.send('session:tokens', {
+          api_token: cookielessSession.cookielessApiToken,
+          api_token_ttl: cookielessSession.cookielessApiTokenTtl,
+          navigation_token: cookielessSession.cookielessNavigationToken,
+          navigation_token_ttl: cookielessSession.cookielessNavigationTokenTtl,
+          session_reference_token_ttl:
+            cookielessSession.cookielessSessionReferenceTokenTtl,
+        })
+
+        // Clear the generateTokensPromise. All IFRAMEs will queue up
+        // on the same promise
+        this._sdk._generateTokensPromise = undefined
+      })
+    }
+    for (const eventType in this._builder.handlers) {
+      for (const handler of this._builder.handlers[eventType]) {
+        this._hostBuilder.on(eventType, (...args) =>
+          handler.apply(this._client, args)
```

Comment:
If the above `session:tokens:request` async handler throws (e.g. when a browser tab is left open for a long enough time that the session token expires), as far as I can see nothing handles this throwing. Could you possibly handle this?

Reply:
If the browser tab is inactive for a period of time exceeding the token time to live, the first thing the Looker application does when the tab becomes active is to request more tokens. The embedding application is expected to get more tokens at that point and return them to the Looker application. If the embed session has not expired the Looker session will continue.